keps/sig-node/00014-resource-api.md

```diff
@@ -0,0 +1,425 @@
+of this proposal.
+
+### Scheduler Extension
+The scheduler needs to watch for NodeStatus ComputeResources field changes and ResourceClass object updates and caches the binding information between the ResourceClass and the matchingComputeResources so that it can serve container resource request expressed through ResourceClass names.
+
+A natural question is how we should define the matching behavior. Suppose there are two ResourceClass objects. ResourceClass RC1 has metadata matching constraint “property1 = value1”, and ResourceClass RC2 has metadata matching constraint “property2 = value2”. Suppose a ComputeResource has both “property1: value1” and “property2: value2” properties, and so match both ResourceClasses. Now should the scheduler consider this ComputeResource as qualified for both ResourceClasses RC1 and RC2, or only one of them?
+
+We feel the desired answer to this question may vary across different types of resources, properties and use cases. To illustrate this, lets consider the following example: A GPU device plugin in a cluster with different types of GPUs may be configured to export a single property, "Type", at the beginning. To support per GPU type resource quota, cluster admins may define the following ResourceClasses:
+
+```yaml
+kind: ResourceClass
+metadata:
+  name: nvidia-k80
+spec:
+  labelSelector:
+    - matchExpressions:
+        - key: "Type"
+          operator: "Eq"
+          values:
+            - "nvidia-tesla-k80"
+
+kind: ResourceClass
+metadata:
+  name: nvidia-p100
+spec:
+  labelSelector:
+    - matchExpressions:
+        - key: "Type"
+          operator: "Eq"
+          values:
+            - "nvidia-tesla-p100"
+```
+
+Later on, suppose the cluster admins add a new GPU node group with a new version of GPU device plugin that exports another resource property "Nvlink" which will be set true for nvidia-tesla-p100 GPUs connected through nvlinks. To utilize this new feature, the cluster admins define the following new ResourceClass with nvlink constraints:
+
+```yaml
+kind: ResourceClass
+metadata:
+  name: nvidia-p100-nvlink
+spec:
+  labelSelector:
+    - matchExpressions:
+        - key: "Type"
+          operator: "Eq"
+          values:
+            - "nvidia-tesla-p100"
+        - key: "Nvlink"
+          operator: "Eq"
+          values:
+            - "true"
+```
+
+Now we face the question that whether the scheduler should allow Pods requesting
+"nvidia-p100" to land on a node in this new GPU node groups. So far, we have
+received different feedbacks on this question. In some use cases, users would
+like to have minimum matching behavior that as long as the underlying hardware
+matches the minimal requirements specified through ResourceClass contraints,
+they want to allow Pods to be scheduled on the hardware. On the other hand, some users desire to reserve expensive hardware resources for users who explicitly request them.
```

Comment:
This seems like the wrong level of granularity. It's way too high. This also seems likely to lead to errors on the cluster admin part as creating these kinds of resource Class requires your cluster admin to have visibility on all the other resource Classes and all the Resources and their properties. 

A smaller granularity, at the node level or resource level would likely be more useful, contained and less error-prone.

Reply:
Not sure I understand your comment. Are you suggesting we shouldn't include `priority` field, or something else?

Reply:
The goal stated here is that you want to have the flexibility to "map `n` Compute Resources to a specific Resource Class".
Alternatively, you can turn it the other way around: "prevent mapping a Compute Resource to a Resource Class" (here a GPU to `nvidia-p100`).

I'm suggesting that the `priority` field is a dangerous way to go if you want to achieve this goal as it is not the right granularity, error prone and the effects of changing one potentially impacts all the resource classes.

Reply:
What we were trying to say here is that different use cases may desire different behaviors. Some users may want minimum matching behavior, i.e., "map n Compute Resources to a specific Resource Class", to hide underlying hardware difference to workloads that don't care about such difference. On the other hand, some users can want to reserve expensive hardware resources for important workloads. Currently people are doing this through taint and toleration, which I think have certain limitations like only works at node level and doesn't integrate well with resource quota. Using priority is similar to using taint in the sense that they may DOS for existing workloads, whether it is intentional or by accident. I do agree that we should provide facility tools to help users use ResourceClass properly. I think @vikaschoudhary16 has included some helpful status fields in ResourceClass to show the ComputeResource matching and allocation information. We will also explore how to display useful information through kubectl. I have been thinking it would be helpful if we can provide some kind of dry-run mode for ResourceClass creation, but not sure whether it is possible to do so through kubectl.

Having said this, I think we are also open to skip this field from the initial implementation. @vikaschoudhary16  and @derekparker do you have any immediate use cases on your side that might depend on the priority field?

Reply:
> to hide underlying hardware difference to workloads that don't care about such difference

Please highlight in the document which workloads and which hardware specifically this requirement is for. Your code is usually already compiled for a specific hardware.

> Currently people are doing this through taint and toleration

I agree that Resource Class need a mechanism to signal that a Compute Resource should belong to a Resource Class. Priority just doesn't seem to be the right model.
It's not an intuitive model, it's completely error prone and dangerous and absolutly not at the right granularity level.

I'd much rather we explore different simpler models:
- Overlapping Resource Class (multiple RC maps to the same Compute Resource) with the ability to remove Comp Res from an RC
- Manually assign CRs to an RC when there is overlap


More generally if we image Resource Class as way to bill my users, I expect them to be able to handle overlap. If I want to see how many GPUs I have then I can just list the Compute Resources.